treescript/src/treescript/github/client.py

```diff
@@ -0,0 +1,137 @@
+"""Treescript git functions."""
+import base64
+import logging
+from collections import defaultdict
+import os
+from string import Template
+from textwrap import dedent
+from typing import Dict, List, Optional, Union
+
+from simple_github import AppClient
+
+log = logging.getLogger(__name__)
+
+GITHUB_APP_ID = 390757  # https://github.com/apps/releng-treescript
+
+
+class GithubClient:
+    def __init__(self, owner, repo):
+        privkey = base64.b64decode(os.environ["SSH_KEY"]).decode("utf-8")
+        self.owner = owner
+        self.repo = repo
+        self._client = AppClient(GITHUB_APP_ID, privkey, owner=owner, repositories=[repo])
+
+    async def close(self):
+        await self._client.close()
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, *excinfo):
+        await self.close()
+
+    async def commit(self, branch: str, message: str, additions: Optional[Dict[str, str]] = None, deletions: Optional[List[str]] = None) -> None:
+        """Commit changes to the given repository and branch.
+
+        Args:
+            branch (str): The branch name to commit to.
+            message (str): The commit message to use.
+            additions (Dict): Files to add or update in the commit. Of the form
+                `{<path>: <contents>}` (optional).
+            deletions (List): Files to delete in the commit. Of the form
+                `[<path>]` (optional).
+        """
+        changes = defaultdict(list)
+        if additions:
+            for name, contents in additions.items():
+                changes["additions"].append({"path": name, "contents": base64.b64encode(contents.encode("utf-8")).decode("utf-8")})
+
+        if deletions:
+            changes["deletions"] = [{"path": p} for p in deletions]
+
+        if not changes:
+            log.warn("No changes to commit, aborting.")
+            return
+
+        query = Template(
+            """
+            query getLatestCommit {
+              repository(owner: "$owner", name: "$repo") {
+                object(expression: "$branch") {
+                  oid
+                }
+              }
+            }
+            """
+        )
+        query = query.substitute(owner=self.owner, repo=self.repo, branch=branch)
+        head_oid = (await self._client.execute(query))["repository"]["object"]["oid"]
+
+        query = """
+            mutation ($input: CreateCommitOnBranchInput!) {
+              createCommitOnBranch(input: $input)  {
+                commit { url }
+              }
+            }
+        """
+        variables = {
+            "input": {
+                "branch": {
+                    "repositoryNameWithOwner": f"{self.owner}/{self.repo}",
+                    "branchName": branch,
+                },
+                "message": {"headline": message},
+                "fileChanges": changes,
+                "expectedHeadOid": head_oid,
```

Comment:
Good question, it doesn't appear to be documented. I'll try sending a bogus oid and see what happens.

Reply:
It returns an error. The odds of a commit landing between these two calls seems very small, but I can add some retry logic.

Reply:
Actually, the `headOid` is just the revision, which is in the task definition. So it's probably better to use that rather than query for it. That way we'd fail rather than adding the bump on top of a new  unexpected commit. Wdyt?

Reply:
On further conversation, we don't want to check the revision because the bump isn't happening on the release branch (at least in the case of our only current use case `firefox-android`).

We may want to add this option later if we ever support a project that doesn't create release branches.. but that's not something to implement now.